### ISSUE TITLE
Processor crashes app after every restart

### ISSUE THREAD

**Issue by kkerley** [ng/paperclip-watermarking-app#2]
It works perfectly in development on my local machine but as soon as I put it into production, it works for the first watermarking, then it crashes every following attempt--both for editing an existing photo and adding new photos.

Here is the error:

Processing PostsController#update (for 76.25.24.165 at 2010-09-21 21:31:24) [PUT]
  Parameters: {"commit"=>"Submit", "post"=>{"photo"=>#File:/tmp/RackMultipart20100921-23361-jbd2x-0, "title"=>"Attack of the shocker, part 7", "url"=>"", "description"=>"Not only is there a shocker, but there's a total random on the left side of the picture.", "status"=>"Pending"}, "action"=>"update", "_method"=>"put", "authenticity_token"=>"MxXDrxfqUivPgcO0lAURhpU9XqFaYQqMcPngi7fM1FY=", "id"=>"16-attack-of-the-shocker-part-7", "controller"=>"posts"}
[paperclip] identify -format %wx%h '/tmp/stream,23361,0.jpg[0]'
[paperclip] convert "/tmp/stream,23361,0.jpg[0]" -resize "300x300>" "/tmp/stream,23361,0,23361,0[0]"

NoMethodError (undefined method `[]' for nil:NilClass):
  lib/paperclip_processors/watermark.rb:10:in`initialize'
  app/controllers/posts_controller.rb:73:in `update'
  app/controllers/posts_controller.rb:72:in`update'
  /dh/passenger/lib/phusion_passenger/rack/request_handler.rb:92:in `process_request'
  /dh/passenger/lib/phusion_passenger/abstract_request_handler.rb:207:in`main_loop'
  /dh/passenger/lib/phusion_passenger/railz/application_spawner.rb:400:in `start_request_handler'
  /dh/passenger/lib/phusion_passenger/railz/application_spawner.rb:351:in`handle_spawn_application'
  /dh/passenger/lib/phusion_passenger/utils.rb:184:in `safe_fork'
  /dh/passenger/lib/phusion_passenger/railz/application_spawner.rb:349:in`handle_spawn_application'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:352:in `__send__'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:352:in`main_loop'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:196:in `start_synchronously'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:163:in`start'
  /dh/passenger/lib/phusion_passenger/railz/application_spawner.rb:209:in `start'
  /dh/passenger/lib/phusion_passenger/spawn_manager.rb:262:in`spawn_rails_application'
  /dh/passenger/lib/phusion_passenger/abstract_server_collection.rb:126:in `lookup_or_add'
  /dh/passenger/lib/phusion_passenger/spawn_manager.rb:256:in`spawn_rails_application'
  /dh/passenger/lib/phusion_passenger/abstract_server_collection.rb:80:in `synchronize'
  /dh/passenger/lib/phusion_passenger/abstract_server_collection.rb:79:in`synchronize'
  /dh/passenger/lib/phusion_passenger/spawn_manager.rb:255:in `spawn_rails_application'
  /dh/passenger/lib/phusion_passenger/spawn_manager.rb:154:in`spawn_application'
  /dh/passenger/lib/phusion_passenger/spawn_manager.rb:287:in `handle_spawn_application'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:352:in`**send**'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:352:in `main_loop'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:196:in`start_synchronously'

Rendering /home/kkerley/iruinphotos.com/public/500.html (500 Internal Server Error)

I've no idea why/how this happens, especially considering it works just fine for a photo before crapping out.  :(

**Comment by jm81** [comment]
I'm having what appears to be the same issue. Looking at the logs, after the first photo, the geometry is not sent correctly, for example, first photo:

```
 /usr/local/bin/composite '-gravity' 'Center' '/u/apps/ms/releases/20101004200017/public/images/watermark.png' '/tmp/stream,13876,4.jpg' '-resize' '500x500>' '/tmp/stream,13876,4,13876,1' 2>/dev/null
```

second photo:

```
 /usr/local/bin/composite '-gravity' 'Center' '/u/apps/ms/releases/20101004200017/public/images/watermark.png' '/tmp/stream,13876,4.jpg' '-resize'  '/tmp/stream,13876,4,13876,1' 2>/dev/null
```

I temporarily fixed by hard coding the scale:

```
  def transformation_command
    scale, crop = @current_geometry.transformation_to(@target_geometry, crop?)
    if watermark_path
      trans = ['-resize', '500x500>']
    else
      trans = %W[-resize #{scale}]
    end
    trans += %W[-crop #{crop} +repage] if crop
    trans << convert_options if convert_options?
    trans
  end
```

I'll try to investigate more later.

Rails 3.0
paperclip 2.3.3
nginx/passenger
ree 1.8.7

**Comment by kkerley** [comment]
I attempted your fix above...no dice.  I even tried the fixed/modified version in the other issue posted here which updated the code for Rails 3 (even though I'm not using Rails 3).  Nothing.  Same stupid error...and yet it works perfectly in development.

This is pretty annoying...especially since it seems to be the ONLY option for watermarking in Rails.  Sigh.  Guess there'll be no watermarking for me in the near future.  I'm glad it's working for you at least.

**Comment by jm81** [comment]
I think it's a problem in paperclip. I've submitted a pull request: http://github.com/thoughtbot/paperclip/pull/333 . For now, adding the following monkey-patch in an initializer would probably work (haven't tested it though):

```
module Paperclip
  class Style
    attr_reader :name, :attachment, :format
    def initialize name, definition, attachment
      @name = name
      @attachment = attachment
      if definition.is_a? Hash
        definition = definition.dup
        @geometry = definition.delete(:geometry)
        @format = definition.delete(:format)
        @processors = definition.delete(:processors)
        @other_args = definition
      else
        @geometry, @format = [definition, nil].flatten[0..1]
        @other_args = {}
      end
      @format = nil if @format.blank?
    end
  end
end
```

**Comment by kkerley** [comment]
Thanks for the response!  I just threw this into an intializer and got the following error when I attempted adding a photo:

ArgumentError in PostsController#update
wrong number of arguments (8 for 3)

I'm still half-asleep so I'm going to hold off on looking at this for a while.

**Comment by exviva** [comment]
@kkerley the Paperclip API apparently changed, so you'll to apply this patch to your watermark processor.

```
diff --git a/lib/paperclip_processors/watermark.rb b/lib/paperclip_processors/watermark.rb
index 3de6e4f..e64120d 100644
--- a/lib/paperclip_processors/watermark.rb
+++ b/lib/paperclip_processors/watermark.rb
@@ -53,7 +53,8 @@ module Paperclip
         end

         begin
-          success = Paperclip.run(command, *params)
+          escaped_params = params.map {|p| "'#{p}'" }
+          success = Paperclip.run(command, escaped_params.join(' '))
         rescue PaperclipCommandLineError
           raise PaperclipError, "There was an error processing the watermark for #{@basename}" if @whiny
         end
```

**Comment by kkerley** [comment]
Thanks for the response and help.  Once again though, it works brilliantly in development and live it crashes the app.  Here's the latest error log for attempting to use it:

ArgumentError (wrong number of arguments (5 for 3)):
  lib/paperclip_processors/watermark.rb:54:in `make'
  app/controllers/posts_controller.rb:53:in`new'
  app/controllers/posts_controller.rb:53:in `create'
  /dh/passenger/lib/phusion_passenger/rack/request_handler.rb:92:in`process_request'
  /dh/passenger/lib/phusion_passenger/abstract_request_handler.rb:207:in `main_loop'
  /dh/passenger/lib/phusion_passenger/railz/application_spawner.rb:400:in`start_request_handler'
  /dh/passenger/lib/phusion_passenger/railz/application_spawner.rb:351:in `handle_spawn_application'
  /dh/passenger/lib/phusion_passenger/utils.rb:184:in`safe_fork'
  /dh/passenger/lib/phusion_passenger/railz/application_spawner.rb:349:in `handle_spawn_application'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:352:in`**send**'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:352:in `main_loop'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:196:in`start_synchronously'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:163:in `start'
  /dh/passenger/lib/phusion_passenger/railz/application_spawner.rb:209:in`start'
  /dh/passenger/lib/phusion_passenger/spawn_manager.rb:262:in `spawn_rails_application'
  /dh/passenger/lib/phusion_passenger/abstract_server_collection.rb:126:in`lookup_or_add'
  /dh/passenger/lib/phusion_passenger/spawn_manager.rb:256:in `spawn_rails_application'
  /dh/passenger/lib/phusion_passenger/abstract_server_collection.rb:80:in`synchronize'
  /dh/passenger/lib/phusion_passenger/abstract_server_collection.rb:79:in `synchronize'
  /dh/passenger/lib/phusion_passenger/spawn_manager.rb:255:in`spawn_rails_application'
  /dh/passenger/lib/phusion_passenger/spawn_manager.rb:154:in `spawn_application'
  /dh/passenger/lib/phusion_passenger/spawn_manager.rb:287:in`handle_spawn_application'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:352:in `__send__'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:352:in`main_loop'
  /dh/passenger/lib/phusion_passenger/abstract_server.rb:196:in `start_synchronously'

**Comment by kkerley** [comment]
I didn't mean to close this and can't figure out how to reopen it

**Comment by exviva** [comment]
Which Paperclip are you using?

**Comment by kkerley** [comment]
It looks like I'm running paperclip 2.3.3.  And I'm on Rails version 2.3.8.  I'm starting to wonder if I need to upgrade my app to Rails 3.0 or downgrade to Rails 2.3.5.

At least the error is different now and it's not passing nil information like before.  For some reason, only 3 arguments are expected but 5 are being passed.  I still don't understand how this happens in production but not in development.  Is there any way it could be Mongrel vs. Passenger?  That doesn't make a lot of sense to me and since I've frozen all gems and everything used in development, deployed it to production and am running that way, I can't imagine it's a problem with a bad gem/plugin combination since they're identical to what I have on dev.

Man, this is frustrating!

**Comment by exviva** [comment]
Maybe your production gemset is mixed up? Rails 3 has all the goodies of Bundler. It looks like your two environments are using two different versions of Paperclip.

**Comment by kkerley** [comment]
How?  Paperclip is in vender/plugins.  It got pushed out with everything else in that folder from development exactly as it is and has never been touched on the production server. :-\

I even did a complete dump this morning from the production server and ran it on my local development machine using exactly what's out on production (apart from Passenger) and it worked perfectly with the watermark processor.  I pushed everything back out, overwriting all files on the production server, restarted, attempted adding a new photo and got the error I posted above.

Hmm...I'm completely out of ideas.  I appreciate your help with everything.  I'll try upgrading to paperclip 2.3.5 tonight after I get home from work and applying your patch to see if that works.  If not, I guess I can see about moving to Rails 3.0 or just not watermarking photos.  At this point, I'm finding it hard to care anymore.
